compiler/rustc_resolve/src/ident.rs

```diff
@@ -1463,9 +1463,28 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                             continue;
                         }
                     }
-                    return PathResult::failed(ident, false, finalize.is_some(), module, || {
-                        ("there are too many leading `super` keywords".to_string(), None)
-                    });
+                    let mut item_type = "module";
+                    let mut suggestion = None;
+                    let label = if path.len() == 1
+                        && let Some(ribs) = ribs
+                        && let RibKind::Normal = ribs[ValueNS][ribs[ValueNS].len() - 1].kind
+                    {
+                        item_type = "item";
+                        suggestion = Some((vec![(ident.span.shrink_to_lo(), "r#".to_string())], "if you still want to call your identifier `super`, use the raw identifier format".to_string(), Applicability::MachineApplicable));
+                        "can't use `super` as an identifier"
+                    } else if segment_idx == 0 {
+                        "can't use `super` on the crate root, there are no further modules to go \"up\" to"
+                    } else {
+                        "there are too many leading `super` keywords"
+                    };
```

Comment:
All this logic should go into the closure, so it's not executed on a good path (`finalize.is_none()`).